File: Data Structures & Algorithms/search-for-word/submission-0.java
class Solution {
    private final int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    public boolean exist(char[][] board, String word) {
        int rows = board.length;
        int cols = board[0].length;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (dfs(board, word, i, j, 0))
                    return true;
            }
        }

        return false;
    }

    private boolean dfs(char[][] board, String word, int row, int col, int index) {
        if (index == word.length())
            return true;

        if (row < 0 || row >= board.length || col < 0 || col >= board[0].length) {
            return false;
        }
        if (board[row][col] != word.charAt(index))
            return false;

        char original = board[row][col];
        board[row][col] = '#';
        for (int[] direction : directions) {
            int nrow = direction[0] + row;
            int ncol = direction[1] + col;
            if (dfs(board, word, nrow, ncol, index + 1)) {
                board[row][col] = original;
                return true;
            }
        }

        board[row][col] = original;
        return false;
    }
}

File: Data Structures & Algorithms/valid-sudoku/submission-0.kt
class Solution {
    fun isValidSudoku(board: Array<CharArray>): Boolean {
        val rows = Array(9){HashSet<Char>()}
        val cols = Array(9){HashSet<Char>()}
        val boxes = Array(9){HashSet<Char>()}

        for(row in 0 until 9){
            for (col in 0 until 9){
                val value = board[row][col]
                if(value=='.') continue
                val boxIndex = (row/3)*3 + (col/3)
                if(!rows[row].add(value)) return false
                if(!cols[col].add(value)) return false
                if(!boxes[boxIndex].add(value)) return false
            }
        }

        return true
    }
}